compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLPasses.td

```diff
@@ -19,6 +19,17 @@ def ROCDLAnnotateKernelForTranslationPass : Pass<
   let dependentDialects = ["ROCDL::ROCDLDialect"];
 }
 
+def ROCDLBufferInstructionsOptimizationPass :
+    InterfacePass<"iree-rocdl-buffer-instructions-optimization", "mlir::FunctionOpInterface"> {
+  let summary = "Optimizations possible with buffer fat pointers";
+  let description = [{
+    Buffer fat pointers support out of bound access so we can make use of this
+    feature for optimizations.
```

Comment:
Can you describe the logic / conditions a bit more, either here or in some comments?

Reply:
Also, to make sure I'm understanding this correctly:

Sometimes (often enough for it to come up?) we hit a case where we're doing a masked load of a vector. With this load, every non-unit dimension of the vector is unmasked. However, some unit dimensions _could_ be masked - if one of those unit dimensions is masked, then the vector is not loaded (or stored I guess).

This pass finds cases that have this form and rewrites them to load + select, taking advantage of oob behavior on buffers.

I think the rewrites make sense, though I'm curious about the context

Reply:
Yes you have the correct understanding.
I have put an example in a gist [here](https://gist.github.com/nirvedhmeshram/2f0f33aeca2c7ab7ede3ba3da2ff747c)
It helps us drop down the use of instructions like `v_cndmask_b32_e64`, this is becuase without the optimization you do elementwise selection for each element of the vector load but with this, you are just doing one select for the whole vector. In the shared exmaple we have 91 cndmask instructions without the optimization and 43 with the optimization. 
Also, the number of lines go from 1630->1406 and sgpr_count 37->20.